src/service/loading-indicator.js

```diff
@@ -0,0 +1,161 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as dom from '../dom';
+import {createLoaderElement} from './loader-element';
+import {htmlFor} from '../static-template';
+import {registerServiceBuilderForDoc} from '../service';
+
+const MIN_SIZE = 20;
+
+/**
+ * @typedef {{
+ *   shown: boolean,
+ *   loader: !Element,
+ *   container: !Element,
+ * }}
+ */
+let LoadingIndicatorStateDef;
+
+/**
+ * @param {!Node|!./ampdoc-impl.AmpDoc} nodeOrDoc
+ */
+export function installLoadingIndicatorForDoc(nodeOrDoc) {
+  registerServiceBuilderForDoc(
+    nodeOrDoc,
+    'loadingIndicator',
+    LoadingIndicatorImpl
+  );
+}
+
+/**
+ * @implements {../service.Disposable}
+ */
+export class LoadingIndicatorImpl {
+  /**
+   * @param {!./ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    /** @private @const */
+    this.ampdoc_ = ampdoc;
+
+    const inViewport = this.inViewport_.bind(this);
+    /** @private @const {!IntersectionObserver} */
+    this.io_ = new ampdoc.win.IntersectionObserver((records) =>
```

Comment:
The current `viewportCallback` is based on a loading rect with a 25% expanded rect. My gut says the safest way to proceed for all of our replacements is to:
1. First recreate as close to the same behavior as possible. In this case it means adding a 25% margin.
2. After landing, if we want, then we can turn up an experiment where we remove the margin and ensure it doesn't negatively affect metrics.

https://github.com/ampproject/amphtml/blob/6faff93057d642e9ecd977faddda85c033d72bf7/src/service/resources-impl.js#L1335-L1340

Reply:
I think you're right here. It was initially expanded given manual measures. But it's probably better to make this consistent for now.